Feature/query parameters

jwir3 opened the pull request:
Adds two methods: `hasQueryParameters()` and `hasQueryParameters(AbstractMap.SimpleEntry<String, String> params...)`. The latter version checks whether the request has _exactly_ the query parameters specified, whereas the former checks whether there are _any_ parameters given at all.

andrzejchm commented on core/src/main/java/io/appflate/restmock/utils/RequestMatchers.java:
```diff
@@ -66,6 +72,55 @@ protected boolean matchesSafely(RecordedRequest item) {
+      };
+    }
+
+    public static RequestMatcher hasQueryParameters(final SimpleEntry<String, String>... expectedParams) {
```
I'd preffer here a dedidacted `QueryParam` class instead of `SimpleEntry` as so the users won't have problems as of what the parameter is all about.

jwir3 replied:
Sure, I can add that.

jwir3 commented:
Similar to #40, I don't seem to be able to reproduce the lint errors locally. I can fix them, but I just need to know what's causing them (I can't see the report(s), either on CircleCI).

jwir3 commented:
As requested, I added a class called `QueryParam`. I also changed the method `hasQueryParameters(QueryParam... expectedParams)` to be named `hasExactQueryParameters(QueryParam... expectedParams)`. I think this makes it more clear that the client needs to match _exactly_ the query parameters given, not just some of them.

andrzejchm commented on core/src/main/java/io/appflate/restmock/utils/RestMockUtils.java:
```diff
@@ -35,4 +42,31 @@ public static MockResponse createResponseFromFile(RESTMockFileParser RESTMockFil
+        final int idx = pair.indexOf("=");
+        final String key = idx > 0 ? URLDecoder.decode(pair.substring(0, idx), "UTF-8") : pair;
+        if (!query_pairs.containsKey(key)) {
+          query_pairs.put(key, new LinkedList<String>());
```
its better to create the list, populate it with items and them put it in a map.

jwir3 replied:
Sure, I'll modify this.

andrzejchm commented on core/src/main/java/io/appflate/restmock/utils/RestMockUtils.java:
```diff
@@ -35,4 +42,31 @@ public static MockResponse createResponseFromFile(RESTMockFileParser RESTMockFil
+        final String key = idx > 0 ? URLDecoder.decode(pair.substring(0, idx), "UTF-8") : pair;
+        if (!query_pairs.containsKey(key)) {
+          query_pairs.put(key, new LinkedList<String>());
+        }
```
if we plan to support multiple values for a single query param, then it would be nice to extend the `QueryParam` class for that by adding a constructor `QueryParam(String key, String... values)` and storing the values in a list. This way the `hasExactQueryParameters` method will really check if the given queryParams EXACTLY matches the ones in the url

jwir3 replied:
Sure, I can do that. I'll make the modifications today and re-push the patch.

jwir3 commented:
@andrzejchm I think I addressed your comments in the latest commits.

jwir3 commented:
Hm, so CircleCI seems to think something is wrong wrt to the FindBugs configuration, but the findbugs.html report in the artifacts doesn't seem to show anything.

jwir3 commented:
Ah. Nevermind. I found the problem. It was complaining about my use of `keySet()` instead of `entrySet()`.

jwir3 commented:
@andrzejchm review ping?

andrzejchm commented on core/src/main/java/io/appflate/restmock/utils/QueryParam.java:
```diff
@@ -0,0 +1,22 @@
+ * A key/value set representing query parameters in an HTTP request.
+ */
+public class QueryParam {
+  private String key;
```
should be final then

andrzejchm commented on core/src/main/java/io/appflate/restmock/utils/RequestMatchers.java:
```diff
@@ -66,6 +70,70 @@ protected boolean matchesSafely(RecordedRequest item) {
+
+            return true;
+          } catch (MalformedURLException e) {
+            return false;
```
additionaly to returning 'false' here it would be nice to add some logs with `RESTMockServer.logger` about  the errors

jwir3 replied:
I can do this if you'd prefer, but the thing is that the only way this could happen is if `item.getPath()` returned a non-unicode string, which seems super unlikely, given that Java `String` objects are unicode by definition.

jwir3 commented:
@andrzejchm Added the logging, as requested. Also I modified `RESTMockServer.logger` to be initialized immediately so that we don't potentially have a race condition where an NPE could crop up (although it's probably super unlikely).
